src/module_5/module5.md

```diff
@@ -0,0 +1,5880 @@
+      <th></th>
+      <th>binary_logloss</th>
+      <th>weighted-return</th>
+      <th>n_trees</th>
+      <th>execution_date</th>
+      <th>diff_ch_sp500_baseline</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>0.759610</td>
+      <td>0.099404</td>
+      <td>39</td>
+      <td>2006-06-30</td>
+      <td>0.049213</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>0.730303</td>
+      <td>0.035528</td>
+      <td>39</td>
+      <td>2006-09-30</td>
+      <td>0.067796</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>0.715078</td>
+      <td>-0.052195</td>
+      <td>39</td>
+      <td>2006-12-30</td>
+      <td>0.068473</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>0.710833</td>
+      <td>-0.067471</td>
+      <td>39</td>
+      <td>2007-03-31</td>
+      <td>0.048029</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>0.711701</td>
+      <td>-0.045395</td>
+      <td>39</td>
+      <td>2007-06-30</td>
+      <td>0.077166</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+
+```python
+(
+    ggplot(
+        test_results_final_tree[test_results_final_tree["weighted-return"] < 10],
+        aes(x="execution_date"),
+    )
+    + geom_point(aes(y="weighted-return"), colour="blue")
+    + geom_point(aes(y="diff_ch_sp500_baseline"), colour="red")
+    + theme(axis_text_x=element_text(angle=90, vjust=0.5, hjust=1))
+)
+```
+
+
+    
+![png](module5_files/module5_46_0.png)
+    
+
+
+
+
+
+    <Figure Size: (640 x 480)>
+
+
+
+
+```python
+print(f"Mean for model: {test_results_final_tree['weighted-return'].mean()}")
+print(
+    f"Mean for baseline: {test_results_final_tree['diff_ch_sp500_baseline'].mean()}\n"
+)
+
+print(f"Median for model: {test_results_final_tree['weighted-return'].median()}")
+print(
+    f"Median for baseline: {test_results_final_tree['diff_ch_sp500_baseline'].median()}\n"
+)
+```
+
+    Mean for model: 4.029023035251342
+    Mean for baseline: 0.022159133577893696
+    
+    Median for model: 0.10471844503860958
+    Median for baseline: 0.015525563344158869
+    
+
+
+Our model does 38 pts better in average and the median is 9 points better, maybe too optimistic to be true. Let's explore what could be happening more deeply:
+
+## 2. Model learning and generalization assessment
+
+In machine learning, the ultimate goal for the model is to try to reduce the error function as much as possible. While this might be positive, sometimes the metric being optimized is not necessarily our best metric based on the business problem we are tackling. In this light, there exists a range of different metrics closer to the model or business, which might be interesting to keep in mind and be diligent to know in which area we are trying to optimize. We can denominate that "pyramid," referring to how close we are with respect to the model, e.g., low-level if we are closer to the model, or high-level if we abstract enough from the model and we are closer to the business.
```

Comment:
Very good explanation. Some people call the "low level" metrics "offline metrics" and the "high level" ones "online metrics"